File: backend/utils/timezone.py
"""
Timezone utilities for Jakarta timezone handling
"""
from datetime import datetime
import pytz

# Jakarta timezone
JAKARTA_TZ = pytz.timezone('Asia/Jakarta')

def get_jakarta_time() -> datetime:
    """Get current time in Jakarta timezone"""
    return datetime.now(JAKARTA_TZ)

def convert_to_jakarta(utc_datetime: datetime) -> datetime:
    """Convert UTC datetime to Jakarta timezone"""
    if utc_datetime.tzinfo is None:
        # If timezone-naive, assume it's UTC
        utc_datetime = pytz.utc.localize(utc_datetime)
    return utc_datetime.astimezone(JAKARTA_TZ)

def convert_to_utc(jakarta_datetime: datetime) -> datetime:
    """Convert Jakarta datetime to UTC"""
    if jakarta_datetime.tzinfo is None:
        # If timezone-naive, assume it's Jakarta time
        jakarta_datetime = JAKARTA_TZ.localize(jakarta_datetime)
    return jakarta_datetime.astimezone(pytz.utc)

def jakarta_now_naive() -> datetime:
    """Get current Jakarta time as timezone-naive datetime (for database)"""
    return get_jakarta_time().replace(tzinfo=None)

def format_jakarta_time(dt: datetime, format_str: str = "%Y-%m-%d %H:%M:%S") -> str:
    """Format datetime in Jakarta timezone"""
    if dt.tzinfo is None:
        # Assume UTC if no timezone info
        dt = pytz.utc.localize(dt)
    jakarta_dt = dt.astimezone(JAKARTA_TZ)
    return jakarta_dt.strftime(format_str)

File: frontend/src/utils/timezone.ts
/**
 * Timezone utilities for Jakarta timezone handling
 */

export const JAKARTA_TIMEZONE = 'Asia/Jakarta';

/**
 * Get current date and time in Jakarta timezone
 */
export const getJakartaTime = (): Date => {
  return new Date().toLocaleString("en-US", { timeZone: JAKARTA_TIMEZONE }) as any;
};

/**
 * Convert UTC date to Jakarta timezone
 */
export const convertToJakarta = (utcDate: Date): Date => {
  return new Date(utcDate.toLocaleString("en-US", { timeZone: JAKARTA_TIMEZONE }));
};

/**
 * Format date in Jakarta timezone
 */
export const formatJakartaTime = (
  date: Date, 
  options: Intl.DateTimeFormatOptions = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false
  }
): string => {
  return date.toLocaleString('id-ID', { 
    ...options, 
    timeZone: JAKARTA_TIMEZONE 
  });
};

/**
 * Get Jakarta time as ISO string
 */
export const getJakartaTimeISO = (): string => {
  const now = new Date();
  const jakartaTime = new Date(now.toLocaleString("en-US", { timeZone: JAKARTA_TIMEZONE }));
  return jakartaTime.toISOString();
};

/**
 * Parse date string and ensure it's in Jakarta timezone
 */
export const parseJakartaTime = (dateString: string): Date => {
  const date = new Date(dateString);
  return convertToJakarta(date);
};
